[MRG] Frontend: edit profile Redux tests

darrenvong opened the pull request:
The first set of tests for the frontend! These tests focus on testing the reducers are correctly updating the Redux store state, rather than via the UI. Gonna hold off UI based tests until we have more stable designs.

Closes #299

darrenvong committed:
test: switch to using createSelectorHook

TypedUseSelectorHook is apparently deprecated

darrenvong committed:
test: stop eager destructure of service in authAction

To break circular dep hell. It works because the thunks don't need to
read the value of service on initial module load

darrenvong committed:
test: use more specific types for profile/hosting pref

to ease testing of not having to set all the things that are unrelated

darrenvong committed:
test: updateUserProfile thunk

darrenvong committed:
test: updateHostingPreference thunk

Also refactored default user into its own helper. Discovered that even
within specific tests, I don't want it to run for every tests!

darrenvong committed:
test: fix proxies so it stops breaking tests

Modified proxy to return the same jest mock functions on property
access. Otherwise we won't ever be able to do any assertion on it
if it returns a value dynamically every time.

Also, only load in jest mocks only in test env. This allows the mocks to
be shared between tests and Storybook. Otherwise, Storybook doesn't seem
to like importing jest and using jest mock functions

darrenvong committed:
test: update imports in stories to new mock location

darrenvong committed:
test: stop story files from importing from mock directly

darrenvong commented on app/frontend/.storybook/main.js:
```diff
@@ -5,13 +5,4 @@ module.exports = {
     "@storybook/addon-essentials",
     "@storybook/preset-create-react-app",
   ],
-  webpackFinal: (config) => {
-    config.resolve.alias["../../service"] = require.resolve(
-      "../src/__mocks__/service.ts"
-    );
-    config.resolve.alias["../../../service"] = require.resolve(
-      "../src/__mocks__/service.ts"
-    );
-    return config;
-  },
```
Had to remove this as this was messing up the stories file import from finding where this mock file is correctly, now that I've moved it.

darrenvong commented on app/frontend/src/features/auth/authActions.ts:
```diff
@@ -2,20 +2,12 @@ import { createAsyncThunk } from "@reduxjs/toolkit";
 import { service } from "../../service";
 import { SignupArguments } from "../../service/user";
 
-const {
-  completeSignup,
-  getCurrentUser,
-  getUser,
-  passwordLogin: apiPasswordLogin,
-  tokenLogin: apiTokenLogin,
-} = service.user;
```
Done to break circular dependency when trying to test. This breaks since `service` hasn't been properly initialised and as this seems to be part of the initial module load, it will try to read from it. Delaying reads from `service` fixes this as by the time the thunk reads from it, it won't be on initial module load and so `service` would have been fully loaded.

amdwit replied:
I think the solution I proposed (about making the service layer ignorant about the model and view) two weeks ago was better. I did that exactly for situations like this. The bad thing about your solution is that this problem will appear more often in the future and that production code is changed for testing purposes.

lucaslcode replied:
If the circular dependencies are this flaky I'm beginning to think we should just do the `setToken` hack and never have to worry about it again.

darrenvong replied:
Do we use the token anywhere else in the app apart from the client needing it? We might be better off moving it out of the store and into something even more global like local storage?

A separate issue, but this has got me thinking that the way we do auth in our app is not that secure, as we're essentially using the token as a session token but it's stored in local storage (with Redux persist, so user is still logged in if they refresh the page) which is known to be vulnerable to XSS. This is [one of](https://developer.okta.com/blog/2018/06/20/what-happens-if-your-jwt-is-stolen) the many nice articles that explains the issue. It mentions JWT but it's not that different from what we're doing right now unless I misunderstood something?

Probably one for a wider architectural discussion though rather than changing it here!

lucaslcode replied:
> Do we use the token anywhere else in the app apart from the client needing it?

Nope, just the interceptors and a few of the API functions need it.

> We might be better off moving it out of the store and into something even more global like local storage?

Seems over-complicated to me. 

I don't know about Local Storage and XSS, but it's just a session token (more like a cookie than a JWT) and the server can invalidate whenever so the JWT problems don't apply as far as I can tell.

darrenvong replied:
> Seems over-complicated to me.

Is it that complicated though if all it takes is `localStorage.setItem('token', <value>)` in the thunk and `localStorage.getItem('token')` to read it wherever you need it? I'd say simpler and less Redux code too as a result, if security turned out to be not a concern. The main friction here is that Redux is global state _for the app_ whereas the clients live outside of the app, especially now with the clean concerns with the service layer.

> it's just a session token (more like a cookie than a JWT) and the server can invalidate whenever

I think the invalidation is also true for JWT, but still a potential attack vector for bad actors in between the time period of when we get around to invalidating the token. Once the user has the token, they're basically free to request as the authenticated user right?

amdwit replied:
Instead of a `setToken` hack we could also initialize the service with a getToken function.

amdwit replied:
instead of `localStorage` we better use `indexedDB`

amdwit replied:
You are right: the token can be used that way. For instance by someone using the same browser later.

amdwit replied:
But for user friendliness it is nice if the user is given the choice to stay logged in.

darrenvong replied:
> instead of `localStorage` we better use `indexedDB`

What would be the benefit of using `indexedDB` if we are not storing a huge amount of data in this case?

lucaslcode replied:
>  Once the user has the token, they're basically free to request as the authenticated user right?

Yes but that's also true for cookies or any system which will keep you logged in.

How about storing the session token in a same site only cookie?

darrenvong replied:
The main difference with cookies is you can have HTTP only cookies, which means JS won't be able to read from it and that eliminates the potential XSS problem. Together with the same site only cookie you mentioned and make sure we only support modern browsers, sounds good to me.

This would require the backend to change to read/set the cookies instead of reading from the auth header though, so maybe best to do it in a separate PR?

darrenvong commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -0,0 +1,117 @@
+import { service as originalService } from "..";
```
Had to move this file here as when we are importing `src/service`, we are actually importing `src/service/index` so the old location didn't work with the jest mocking.

darrenvong commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -0,0 +1,117 @@
+} as unknown) as typeof originalService;
+
+export const service = new Proxy(
+  {},
```
Most of the changes here are needed to ensure the proxy returns the same mock every time when accessed via the proxy. Otherwise, this was screwing up the tests since there'd have been no way to retrieve the original mock that has been called (it was returning a new function on the fly each time!) that I need to perform test assertions on.

darrenvong commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -0,0 +1,117 @@
+              };
+
+              const proxiedMockFn =
+                process.env.NODE_ENV === "test"
```
Had to do this otherwise we can't use this in Storybook. It didn't like trying to load `jest` in...

darrenvong commented on app/frontend/src/service/user.ts:
```diff
@@ -20,11 +20,30 @@ import { ProtoToJsTypes } from "../utils/types";
   >;
 };
+
+export type UpdateUserProfileData = Pick<
+  ProfileFormData,
+  | "name"
+  | "city"
+  | "gender"
+  | "occupation"
+  | "languages"
+  | "aboutMe"
+  | "aboutPlace"
+  | "countriesVisited"
+  | "countriesLived"
+>;
+
+export type HostingPreferenceData = Omit<
+  ProfileFormData,
+  keyof UpdateUserProfileData | "color" | "hostingStatus"
+>;
```
Made these new two types to make testing easier without type asserting too much, otherwise I'd have had to set a lot of unrelated fields. It's probably more accurate types for the respective form code anyway.

darrenvong commented on app/frontend/src/store.ts:
```diff
@@ -59,6 +59,6 @@ export type AppDispatch = typeof store.dispatch;
 
 export const useAppDispatch = () => useDispatch<AppDispatch>();
-export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector;
+export const useTypedSelector = createSelectorHook<RootState>();
```
Switched to using `createSelectorHook` as I was told by the editor `TypedUseSelectorHook` is deprecated.

amdwit commented on app/frontend/src/features/profile/actions.test.ts:
```diff
@@ -0,0 +1,179 @@
+  .updateHostingPreference as jest.Mock;
+
+describe("updateUserProfile thunk", () => {
+  test("updates the store with the latest user profile info", async () => {
```
What about using `it` for readability? Together with the first argument it then forms a full sentence.
```suggestion
  it("updates the store with the latest user profile info", async () => {
```

darrenvong replied:
I used `test` as that was the advertised API, but I see they also have `it` as an alias. Yeah let's do it, more readable and less typing!

amdwit commented:
Are we using the same mock both for Storybook and for Jest? Is that a good idea? Storybook is for demoing and some human testing. Jest testing may break ci/cd.

darrenvong commented:
> Are we using the same mock both for Storybook and for Jest? Is that a good idea? Storybook is for demoing and some human testing. Jest testing may break ci/cd.

I thought that was your initial intention when you put `service.ts` in a `__mocks__` folder. But now you've asked and got me thinking, I'd say probably not, as the Storybook file shouldn't contain stuff that needs to call the API?

In that case, it might be better to put the shared stuff meant only for Storybook somewhere else that's not in `__mocks__` folders, as they have a special meaning to jest for defining manual mocks.

darrenvong committed:
test: switch to using "it" over "test"

Tidied up service manual mock to only include Jest stuff as it shoudn't
be used outside of Jest tests.

amdwit commented:
But Jest tests should not call the API too? A separate mock folder for Storybook, seems the best solution to me.

darrenvong commented:
Sounds good to me. We already have a `stories` folder so maybe the Storybook specific mocks can go there if they need to be shared between story files?

As it is, I pulled out the static data into fixture JSON files under `test`. Since they are static, I think they are probably safe to be shared between test and story files. Don't really have a strong preference with this either way.

amdwit commented:
Let's see how it goes then. I just fear that some changes for storybook lead to broken unit tests.

darrenvong commented:
I just removed the conditional logic in the mock file for the proxy to always return jest functions, which means it should only be used in jest tests. So let's create separate Storybook mocks in future like you suggested if we need them in the story files.

darrenvong commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -0,0 +1,101 @@
+              methodName
+            ];
+            if (serviceMethod) {
+              const proxiedMockFn = jest.fn((...args: any[]) => {
```
I removed the conditional logic to simplify this and this now always return jest functions, which means it should only be used in jest tests. So let's create separate Storybook mocks in future if we need them in the story files.

darrenvong committed:
test: value setting logic in service calls related to profile

darrenvong commented on app/frontend/src/service/user.test.ts:
```diff
@@ -0,0 +1,128 @@
+import { Empty } from "google-protobuf/google/protobuf/empty_pb";
```
Added some more unit tests on the service layer, mainly to check the values are set correctly before it gets sent to the grpc client, which is still mocked. I thought this is still worthwhile as there's still a little bit of logic in these and weren't tested in the reducer tests as this layer has been mocked out in those.

amdwit reviewed (approved):
Unless you go for 100% coverage 🤣
